drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c

```diff
@@ -0,0 +1,279 @@
+ * NOTE: The PWM output maximum is 324Hz in EC LPM, so if we need fan to work
+ *       then don't let EC enter LPM.
+ */
+#define TACH_FREQ		EC_FREQ
+#define IT8XXX2_PWM_T0_MASK	(IT8XXX2_PWM_T0DVS | IT8XXX2_PWM_T0CHSEL)
+#define IT8XXX2_PWM_T1_MASK	(IT8XXX2_PWM_T1DVS | IT8XXX2_PWM_T1CHSEL)
+#define GPD6_PINCTRL_IDX	0
+#define GPJ2_PINCTRL_IDX	1
+#define GPD7_PINCTRL_IDX	0
+#define GPJ3_PINCTRL_IDX	1
+
+/* Device config */
+struct tach_alt_cfg {
+	/* Pinmux control device structure */
+	const struct device *pinctrls;
+	/* GPIO pin */
+	uint8_t pin;
+	/* Alternate function */
+	uint8_t alt_fun;
+};
+
+struct tach_it8xxx2_config {
+	/* Fan x tachometer LSB reading register */
+	uintptr_t reg_fxtlrr;
+	/* Fan x tachometer MSB reading register */
+	uintptr_t reg_fxtmrr;
+	/* Tachometer switch control register */
+	uintptr_t reg_tswctlr;
+	/* Tachometer alternate configuration list */
+	const struct tach_alt_cfg *alt_list;
+	/* Select channel of tachometer */
+	int channel;
+	/* Number of pulses per round of tachometer's input */
+	int pulses_per_round;
+};
+
+/* Driver data */
+struct tach_it8xxx2_data {
+	/* Captured counts of tachometer */
+	uint32_t capture;
+};
+
+/* Driver convenience defines */
+#define DRV_CONFIG(dev)	((const struct tach_it8xxx2_config *)(dev)->config)
+#define DRV_DATA(dev)	((struct tach_it8xxx2_data *)(dev)->data)
+
+static int tach_ch_valid(const struct device *dev, int tach_ch)
+{
+	const struct tach_it8xxx2_config *const config = DRV_CONFIG(dev);
+	volatile uint8_t *reg_tswctlr = (uint8_t *)config->reg_tswctlr;
+	int valid = 0;
+
+	switch (tach_ch) {
+	case TACH0_CHANNEL_A:
+		if ((*reg_tswctlr & IT8XXX2_PWM_T0_MASK) == IT8XXX2_PWM_T0DVS)
+			valid = 1;
+		break;
+	case TACH0_CHANNEL_B:
+		if ((*reg_tswctlr & IT8XXX2_PWM_T0_MASK) == IT8XXX2_PWM_T0_MASK)
+			valid = 1;
+		break;
+	case TACH1_CHANNEL_A:
+		if ((*reg_tswctlr & IT8XXX2_PWM_T1_MASK) == IT8XXX2_PWM_T1DVS)
+			valid = 1;
+		break;
+	case TACH1_CHANNEL_B:
+		if ((*reg_tswctlr & IT8XXX2_PWM_T1_MASK) == IT8XXX2_PWM_T1_MASK)
+			valid = 1;
+		break;
+	default:
+		break;
+	}
+
+	return valid;
+}
+
+int tach_it8xxx2_sample_fetch(const struct device *dev,
+			      enum sensor_channel chan)
+{
+	const struct tach_it8xxx2_config *const config = DRV_CONFIG(dev);
+	volatile uint8_t *reg_fxtlrr = (uint8_t *)config->reg_fxtlrr;
+	volatile uint8_t *reg_fxtmrr = (uint8_t *)config->reg_fxtmrr;
+	volatile uint8_t *reg_tswctlr = (uint8_t *)config->reg_tswctlr;
+	int tach_ch = config->channel;
+	struct tach_it8xxx2_data *const data = DRV_DATA(dev);
+
+	if (chan != SENSOR_CHAN_RPM) {
+		LOG_ERR("Sensor chan %d, only support SENSOR_CHAN_RPM", chan);
+		return -ENOTSUP;
+	}
+
+	if (tach_ch_valid(dev, tach_ch)) {
+		/* If channel data of tachometer is valid, then save it */
+		data->capture = ((*reg_fxtmrr) << 8) | (*reg_fxtlrr);
+
+		if ((tach_ch == TACH0_CHANNEL_A) || (tach_ch == TACH0_CHANNEL_B))
+			/* Clear tachometer0 data valid status */
+			*reg_tswctlr |= IT8XXX2_PWM_T0DVS;
```

Comment:
Please add braces even for single-line code blocks. See https://docs.zephyrproject.org/latest/contribute/index.html#coding-style

Reply:
Done